### ISSUE TITLE
"'NoneType' object is not callable" on program close when using local data file

### ISSUE THREAD

**Issue by sjpritchard** [qdrant/qdrant-client#365]
Description:

When using a local data file, the following exception is generated on program termination (with no apparent adverse effects on the data):

```
Exception ignored in: <function QdrantClient.__del__ at 0x7f0eb98f1e40>
Traceback (most recent call last):
  File "/xxxx/lib/python3.11/site-packages/qdrant_client/qdrant_client.py", line 116, in __del__
  File "/xxxx/lib/python3.11/site-packages/qdrant_client/qdrant_client.py", line 123, in close
  File "/xxxx/lib/python3.11/site-packages/qdrant_client/local/qdrant_local.py", line 57, in close
TypeError: 'NoneType' object is not callable
```

Code sample:
```
qdrant_client = QdrantClient(path="./qdrant_data")
qdrant_client.recreate_collection(
    collection_name="xxxxx",
    vectors_config=models.VectorParams(size=1024, distance=models.Distance.COSINE),
)
for idx, item in enumerate(input_data, start=1):
    print(f"Processing {idx} of {len(input_data)}")
    text_string = " ".join(item["text"])
    if not text_string:
        continue
    embedding = get_embedding(text_string)
    id = str(uuid.uuid4())
    qdrant_client.upsert(
        collection_name="simcortex", points=[models.PointStruct(id=id, vector=embedding, payload={"text": text_string})]
    )
```

System Info:
- Python 3.11
- Ubuntu 22.04 on WSL
- qdrant-client 1.6.4

**Comment by generall** [comment]
could you please provide a code example of how to reproduce the issue? I am afraid, your explanation leaves a room for misinterpreting

**Comment by joein** [comment]
It seems like it might be related to the usage of `portalocker` on WSL. At least that's what can be assumed from the log.

**Comment by sjpritchard** [comment]
> could you please provide a code example of how to reproduce the issue? I am afraid, your explanation leaves a room for misinterpreting

I've added some example code. It's pretty simple. Create a client, and multiple upserts. I tried adding an explicit qdrant_client.close() at the end with no effect.

**Comment by XXXHlown** [comment]
Has this issue been resolved? I got the similar Exception while data upsert

Code sample:
``` python
class Model(object):
    """singleton & lazy load"""

    def __init__(self) -> None:
        self._qclient = None
        self._hf_embeddings = None

    @property
    def qclient(self):
        if not self._qclient:
            self._qclient = qdrant_client.QdrantClient(path="xxx")
        return self._qclient

    @property
    def hf_embeddings(self):
        if not self._hf_embeddings:
            self._hf_embeddings = HuggingFaceEmbeddings(model_name="xxx")
        return self._hf_embeddings


MODEL = Model()


@utils.time_warpper
def upsert_qdrant(points: List[PointStruct]):
    col_name = utils.get_qdrant_col_name()
    MODEL.qclient.upsert(col_name, points)


points = []
for i in range(100):
    text = f"hello {i}"
    p = PointStruct(
        vector=MODEL.hf_embeddings.encode(text),
        payload={
            "text": text,
        },
        ids=[str(i)],
    )
    points.append(p)

upsert_qdrant(points)
```

result:
``` sh
Exception ignored in: <function QdrantClient.__del__ at 0x7fe43185ec10>
Traceback (most recent call last):
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/qdrant_client.py", line 116, in __del__
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/qdrant_client.py", line 123, in close
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/local/qdrant_local.py", line 57, in close
```


sys info:
```
env: conda python 3.9
qdrant-client: 1.6.3
OS: CentOS Linux release 7.9.2009
```

**Comment by joein** [comment]
Hi @XXXHlown @UnlcleTian, sorry, I didn't see your comments, thanks for the code sample, I will try to reproduce it and come back later this week

**Comment by XXXHlown** [comment]
@joein 
Hello, thanks for your patient for this BUG.
My local collection `meta.json`'s content is:
``` json
{
    "collections": {
        "nengyuan": {
            "vectors": {
                "content": {
                    "size": 768,
                    "distance": "Cosine",
                    "hnsw_config": null,
                    "quantization_config": null,
                    "on_disk": null
                }
            },
            "shard_number": null,
            "replication_factor": null,
            "write_consistency_factor": null,
            "on_disk_payload": null,
            "hnsw_config": null,
            "wal_config": null,
            "optimizers_config": null,
            "init_from": null,
            "quantization_config": null
        }
    },
    "aliases": {}
}
```
I try to add debug info as u provided code like bellow:
``` python
class QdrantLocal(QdrantBase):
    ...
    def close(self, **kwargs: Any) -> None:
        self._closed = True
        print(f'{self.collections=}')
        for collection in self.collections.values():
            print(f'for each delete, {collection=}')
            collection.close()
        if self._flock_file is not None:
            portalocker.unlock(self._flock_file)
```
And here is the output:
``` sh
# output for above
self.collections={'nengyuan': <qdrant_client.local.local_collection.LocalCollection object at 0x7fb9b92f6ca0>}
for each delete, collection=<qdrant_client.local.local_collection.LocalCollection object at 0x7fb9b92f6ca0>
Exception ignored in: <function QdrantClient.__del__ at 0x7fb9ba33f280>
Traceback (most recent call last):
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/qdrant_client.py", line 116, in __del__
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/qdrant_client.py", line 123, in close
  File "/home/xxx/.conda/envs/xxx_py39/lib/python3.9/site-packages/qdrant_client/local/qdrant_local.py", line 59, in close
TypeError: 'NoneType' object is not callable
```
It seems like `portalocker.unlock(self._flock_file)` is the main reason.
Thanks a lot.

**Comment by joein** [comment]
Hi @XXXHlown , sorry for the late response and thank you for the debug info!

Maybe you could also provide `portalocker` version?

Have you tried to reinstall portalocker?

**Comment by joein** [comment]
I was finally able to reproduce it on my own

**Comment by joein** [comment]
It has to be fixed with #421 

It will be available with the next version

**Comment by joein** [comment]
I am closing it for now, feel free to open if the issue persists with the next version (it will be 1.7.1)
